libraries/AP_Torqeedo/AP_Torqeedo_TQBus.cpp

```diff
@@ -0,0 +1,1119 @@
+            }
+        }
+    }
+
+    // calculate upper and lower limits for forward-reverse transition delay
+    int16_t lower_limit = -1000;
+    int16_t upper_limit = 1000;
+    if (_dir_limit < 0) {
+        upper_limit = 0;
+    }
+    if (_dir_limit > 0) {
+        lower_limit = 0;
+    }
+
+    // calculate dt since last update
+    float dt = (now_ms - _motor_speed_limited_ms) * 0.001f;
+    if (dt > 1.0) {
+        // after a long delay limit motor output to zero to avoid sudden starts
+        lower_limit = 0;
+        upper_limit = 0;
+    }
+    _motor_speed_limited_ms = now_ms;
+
+    // apply slew limit
+    if (_slew_time > 0) {
+       const float chg_max = 1000.0 * dt / _slew_time;
+       _motor_speed_limited = constrain_float(desired_motor_speed, _motor_speed_limited - chg_max, _motor_speed_limited + chg_max);
+    } else {
+        // no slew limit
+        _motor_speed_limited = desired_motor_speed;
+    }
+
+    // apply upper and lower limits
+    _motor_speed_limited = constrain_float(_motor_speed_limited, lower_limit, upper_limit);
+
+    // record time motor speed becomes zero
+    if (is_zero(_motor_speed_limited)) {
+        if (_motor_speed_zero_ms == 0) {
+            _motor_speed_zero_ms = now_ms;
+        }
+    } else {
+        // clear timer
+        _motor_speed_zero_ms = 0;
+    }
+
+    return (int16_t)_motor_speed_limited;
+}
+
+// output logging and debug messages (if required)
+// force_logging should be true if caller wants to ensure the latest status is logged
+void AP_Torqeedo_TQBus::log_TRQD(bool force_logging)
+{
+    // exit immediately if options are all unset
+    if (_options == 0) {
+        return;
+    }
+
+    // return if not enough time has passed since last output
+    const uint32_t now_ms = AP_HAL::millis();
+    if (!force_logging && (now_ms - _last_log_TRQD_ms < TORQEEDO_LOG_TRQD_INTERVAL_MS)) {
+        return;
+    }
+    _last_log_TRQD_ms = now_ms;
+
+    const bool health = healthy();
+    int16_t actual_motor_speed = get_motor_speed_limited();
+
+    if ((_options & options::LOG) != 0) {
+        // @LoggerMessage: TRQD
+        // @Description: Torqeedo Status
+        // @Field: TimeUS: Time since system startup
+        // @Field: Health: Health
+        // @Field: DesMotSpeed: Desired Motor Speed (-1000 to 1000)
+        // @Field: MotSpeed: Motor Speed (-1000 to 1000)
+        // @Field: SuccCnt: Success Count
+        // @Field: ErrCnt: Error Count
+        AP::logger().Write("TRQD", "TimeUS,Health,DesMotSpeed,MotSpeed,SuccCnt,ErrCnt", "QBhhII",
+                           AP_HAL::micros64(),
+                           health,
+                           _motor_speed_desired,
+                           actual_motor_speed,
+                           _parse_success_count,
+                           _parse_error_count);
+    }
+
+    if ((_options & options::DEBUG_TO_GCS) != 0) {
+        gcs().send_text(MAV_SEVERITY_INFO,"TRQD h:%u dspd:%d spd:%d succ:%ld err:%ld",
```

Comment:
we might want to add the instance number in this debug output as well although it's less important because it's just debug.